source/open_node/ast.py

```diff
@@ -0,0 +1,1017 @@
+
+
+class Statement(OpenNode):
+    """A statement: anything that can appear on its own line"""
+
+
+@dataclass
+class AliasStatement(Statement):
+    """
+    Alias statement
+
+    Example::
+
+        let a = qubits[0];
+
+    """
+
+    target: Identifier
+    value: Union[IndexIdentifier, Identifier]
+
+
+@dataclass
+class ExpressionStatement(Statement):
+    """A statement that contains a single expression"""
+
+    expression: Expression
+
+
+class QubitDeclTypeName(Enum):
+    qubit = auto()  # OpenQASM 3 style.
+    qreg = auto()  # OpenQASM 2 style.
+
+
+@dataclass
+class QubitDeclaration(Statement):
+    """
+    Global qubit declaration
+
+    Example::
+
+        qubit q;
+        qubit[4] q;
+
+        q // <- qubit
+        [4] // <- designator
+
+    """
+
+    decl_type_name: QubitDeclTypeName
+    qubit: Qubit
+    designator: Optional[Expression]
+
+
+@dataclass
+class QuantumGateDefinition(Statement):
+    """
+    Define a new quantum gate
+
+    Example::
+
+        gate cx c, t {
+            ctrl @ unitary(pi, 0, pi) c, t;
+        }
+
+    """
+
+    name: str
+    arguments: List[ClassicalArgument]
+    qubits: List[Identifier]
+    body: List[QuantumStatement]
+
+
+class QuantumStatement(Statement):
+    """Statements that may appear inside a gate declaration"""
+
+
+@dataclass
+class QuantumForInLoop(Statement):
+    """For In loops that contain only quantum statements."""
+
+    loop_variable: Identifier
+    set_declaration: Union[RangeDefinition, List[Expression], Identifier]
+    block: List[QuantumStatement]
+
+
+@dataclass
+class QuantumWhileLoop(Statement):
+    """While loops that contain only quantum statements."""
+
+    while_condition: Expression
+    block: List[QuantumStatement]
+
+
+@dataclass
+class ExternDeclaration(Statement):
+    """
+    A extern declaration
+
+    Example::
+
+        extern get_pauli(int[prec]) -> bit[2 * n];
+
+        get_pauli  // <- name
+        int[prec]  // <- classical type
+        bit[2 * n] // <- return type
+
+    """
+
+    name: str
+    classical_types: List[ClassicalType]
+    return_type: Optional[ClassicalType]
+
+
+@dataclass
+class Qubit(OpenNode):
+    """
+    A qubit
+
+    Example::
+
+        qubit q;
+
+        q  // <- Qubit
+
+    """
+
+    name: str
+
+
+class Expression(OpenNode):
+    """An expression: anything that returns a value"""
+
+
+UnaryOperator = Enum("UnaryOperator", "~ ! -")
+
+
+@dataclass
+class UnaryExpression(Expression):
+    """
+    A unary expression
+
+    Example::
+
+        ~b
+        !bool
+        -i
+
+    """
+
+    op: UnaryOperator
+    expression: Expression
+
+
+BinaryOperator = Enum("BinaryOperator", "> < >= <= == != && || | ^ & << >> + - * / % **")
+
+
+@dataclass
+class BinaryExpression(Expression):
+    """
+    A binary expression
+
+    Example::
+
+        q1 || q2
+
+    """
+
+    op: BinaryOperator
+    lhs: Expression
+    rhs: Expression
+
+
+ConstantName = Enum("ConstantName", "pi π tau 𝜏 euler ℇ")
```

Comment:
From the AST's perspective I think `pi` and `π` are synonyms; different tokens that represent the same thing. Therefore this enum should just be `Enum("ConstantName", "pi tau euler")`

Reply:
+1

Reply:
If we remove π, 𝜏, ℇ from the enum, we will no know from the AST whether we have pi or π in the code. When we print the AST, both pi and π will be printed as pi.

If we want to preserve printing, we can add another str field to keep the information.

What do you think?

Reply:
I don't think we should preserve this - similar to how we don't preserve comments, spacing, and other things that don't affect the semantics of the program.

Reply:
One thought which may or may not be useful:
My (two) hacked-together ASTs for OpenQASM 2 [[1](https://github.com/jwoehr/nuqasm2),[2](https://github.com/Qiskit/qiskit-terra/pull/6657)] preserve source lines for "reasons". No harm in that, the AST can be dumped/saved  for debugging, etc., allowing the programmer recognizes their own code when it appears in the debugger or reconstruction/retreival, or discarded if not needed.